### ISSUE TITLE
improve `unresolved-reference` for cases where a variable is declared but unbound

### ISSUE THREAD

**Issue by KotlinIsland** [astral-sh/ty#2665]
### Summary

```py
from typing import Literal

def bool_or_cake() -> bool | Literal["🎂"]:
    return bool()  # or cake?

a: int

print(a)  # Variable `a` is unbound / uninitialised (unbound-name)

b: int

if bool_or_cake():
    b = 1

print(b)  # Variable `b` may be unbound / uninitialised (possibly-unbound-name)

print(c)  # Name `c` used when not defined (unresolved-reference)
```
[playground](https://play.ty.dev/d07cdd97-0d0b-4394-953d-d10b9e596889)

for `a` and `b` these are most definitely "resolvable" and "defined", so this error message is quite perplexing. i've included suggestions for potential replacements

for `possibly-unresolved-reference` this name is really strange, i think it should definitely be changed

**Comment by carljm** [comment]
I think [this version](https://play.ty.dev/5426c9cc-5f5e-4129-8666-4c1c06061726) of the repro is more useful:

```py
def coinflip() -> bool:
    return True

a: int

print(a)

b: int

if coinflip():
    b = 1

print(b)

print(c)
```

ty knows that `bool()` always returns `False`, so it considers the `if bool():` branch unreachable, meaning that the `a` and `b` cases in your version are actually identical. In my version we don't error on `b` at all by default; the `possibly-unresolved-reference` diagnostic is disabled by default because there are too many cases where it can have false positives.

But I take the core point here to be that in our diagnostics we should distinguish "declared but never bound" from "neither declared nor bound", and I do think this makes sense. Thanks for the report.

**Comment by KotlinIsland** [comment]
> ty knows that `bool()` always returns `False`

ah yes, I forgot about that. makes minifying very annoying hahaha

why isn't there a diagnostic to report that the condition is constant?

**Comment by KotlinIsland** [comment]
> `possibly-unresolved-reference`

this name is really strange haha, i think it should definetly be changed
